src/modm/processing/fiber/functions.hpp

```diff
@@ -50,29 +51,85 @@ get_id()
 }
 
 /**
- * Yields the current fiber until the time interval has elapsed.
- * This functionality is a convenience wrapper around `modm::Timeout`
- * if interval ≥1ms or `modm::PreciseTimeout` if interval ≥1µs.
- * For nanosecond delays, use `modm::delay(ns)`.
+ * Yields the current fiber until `condition()` returns true or the time
+ * duration has elapsed.
+ *
+ * @returns `true` if the condition was met, `false` if the time duration has
+ *          elapsed.
  *
  * @note Due to the overhead of `yield()` and the scheduling other fibers, the
- *       sleep interval may be longer without any guarantee of an upper limit.
+ *       sleep duration may be longer without any guarantee of an upper limit.
  */
-template< typename Rep, typename Period >
-void
-sleep_for(std::chrono::duration<Rep, Period> interval)
+template< class Rep, class Period, class Function >
+[[nodiscard]] bool
+poll_for(std::chrono::duration<Rep, Period> sleep_duration, Function &&condition)
 {
 	// Only choose the microsecond clock if necessary
-	using TimeoutType = std::conditional_t<
+	using Clock = std::conditional_t<
 		std::is_convertible_v<std::chrono::duration<Rep, Period>,
 							  std::chrono::duration<Rep, std::milli>>,
-		modm::GenericTimeout< modm::chrono::milli_clock, modm::chrono::milli_clock::duration>,
-		modm::GenericTimeout< modm::chrono::micro_clock, modm::chrono::micro_clock::duration>
-	>;
+		modm::chrono::milli_clock, modm::chrono::micro_clock>;
+
+	const auto start = Clock::now();
+	do {
+		if (condition()) return true;
+		modm::this_fiber::yield();
+	}
+	while((Clock::now() - start) <= sleep_duration);
+	return false;
+}
+
+/**
+ * Yields the current fiber until `condition()` returns true or the sleep time
+ * has been reached.
+ *
+ * @returns `true` if the condition was met, `false` if the sleep time has
+ *          elapsed.
+ *
+ * @note Due to the overhead of `yield()` and the scheduling other fibers, the
+ *       sleep duration may be longer without any guarantee of an upper limit.
+ */
+template< class Clock, class Duration, class Function  >
+[[nodiscard]] bool
+poll_until(std::chrono::time_point<Clock, Duration> sleep_time, Function &&condition)
+{
+	const auto start = Clock::now();
+	const auto sleep_duration = sleep_time - start;
+	do {
+		if (condition()) return true;
```

Comment:
```suggestion
		if (std::forward<Function>(condition)()) return true;
```

to be 100% generic. For instance, if `condition` binds to an r-value and its type has an `operator()` with a `&&` ref-qualifier it won't get called without the `std::forward`. Probably never an issue in practice but still more correct :wink:.